# Optimizing Data Acquisition

Since this is a pretty extensive dataset, I'll explore jsut one subset of the data. Let's see what can be done to reduce size

In [49]:
import pandas as pd
ny = pd.read_csv(r'..\data\raw\taxi data\yellow_tripdata_2015-01_00', parse_dates=['tpep_pickup_datetime', 'tpep_dropoff_datetime'])
ny.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1         1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2         1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3         1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4         1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude  RateCodeID  \
0           1.59        -73.993896        40.750111           1   
1           3.30        -74.001648        40.724243           1   
2           1.80        -73.963341        40.802788           1   
3           0.50        -74.009087        40.713818           1   
4           3.00        -73.971176        40.762428           1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -73.974785         40.750618             1   
1                  N         -73.994415         40.759109             1   
2                  N         -73.951820         40.824413             2   
3                  N         -74.004326         40.719986             2   
4                  N         -74.004181         40.742653             2   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0         12.0    1.0      0.5        3.25           0.0   
1         14.5    0.5      0.5        2.00           0.0   
2          9.5    0.5      0.5        0.00           0.0   
3          3.5    0.5      0.5        0.00           0.0   
4         15.0    0.5      0.5        0.00           0.0   

   improvement_surcharge  total_amount  
0                    0.3         17.05  
1                    0.3         17.80  
2                    0.3         10.80  
3                    0.3          4.80  
4                    0.3         16.30

In [50]:
ny.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499999 entries, 0 to 499998
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               499999 non-null  int64         
 1   tpep_pickup_datetime   499999 non-null  datetime64[ns]
 2   tpep_dropoff_datetime  499999 non-null  datetime64[ns]
 3   passenger_count        499999 non-null  int64         
 4   trip_distance          499999 non-null  float64       
 5   pickup_longitude       499999 non-null  float64       
 6   pickup_latitude        499999 non-null  float64       
 7   RateCodeID             499999 non-null  int64         
 8   store_and_fwd_flag     499999 non-null  object        
 9   dropoff_longitude      499999 non-null  float64       
 10  dropoff_latitude       499999 non-null  float64       
 11  payment_type           499999 non-null  int64         
 12  fare_amount            499999 non-null  floa

I'll try to reduce the footprint by selecting the data types used for each field.

In [51]:
dtype = {
    'VendorID': 'category',
    'passenger_count': 'int8',
    'trip_distance': 'float32',
    'pickup_longitude': 'float32',
    'pickup_latitude': 'float32',
    'dropoff_longitude': 'float32',
    'dropoff_latitude': 'float32',
    'RateCodeID': 'category',
    'store_and_fwd_flag': 'category',
    'payment_type': 'category',
    'fare_amount': 'float32',
    'extra': 'float32',
    'mta_tax': 'category',
    'tip_amount': 'float32',
    'tolls_amount': 'float32',
    'improvement_surcharge': 'category',
    'total_amount': 'float32'
}

ny = pd.read_csv(r'..\data\raw\taxi data\yellow_tripdata_2015-01_00', dtype=dtype, parse_dates=['tpep_pickup_datetime', 'tpep_dropoff_datetime'])
ny.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499999 entries, 0 to 499998
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   VendorID               499999 non-null  category      
 1   tpep_pickup_datetime   499999 non-null  datetime64[ns]
 2   tpep_dropoff_datetime  499999 non-null  datetime64[ns]
 3   passenger_count        499999 non-null  int8          
 4   trip_distance          499999 non-null  float32       
 5   pickup_longitude       499999 non-null  float32       
 6   pickup_latitude        499999 non-null  float32       
 7   RateCodeID             499999 non-null  category      
 8   store_and_fwd_flag     499999 non-null  category      
 9   dropoff_longitude      499999 non-null  float32       
 10  dropoff_latitude       499999 non-null  float32       
 11  payment_type           499999 non-null  category      
 12  fare_amount            499999 non-null  floa

We have gone from 72.5 MB to to 28.1 MB.

# Cleaning Data

Now we'll try to find inconsistencies in the data. Let's look at its description first.

In [52]:
ny.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0        2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1        1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2        1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3        1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4        1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
0           1.59        -73.993896        40.750111          1   
1           3.30        -74.001648        40.724243          1   
2           1.80        -73.963341        40.802788          1   
3           0.50        -74.009087        40.713818          1   
4           3.00        -73.971176        40.762428          1   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
0                  N         -73.974785         40.750618            1   
1                  N         -73.994415         40.759109            1   
2                  N         -73.951820         40.824413            2   
3                  N         -74.004326         40.719986            2   
4                  N         -74.004181         40.742653            2   

   fare_amount  extra mta_tax  tip_amount  tolls_amount improvement_surcharge  \
0         12.0    1.0     0.5        3.25           0.0                   0.3   
1         14.5    0.5     0.5        2.00           0.0                   0.3   
2          9.5    0.5     0.5        0.00           0.0                   0.3   
3          3.5    0.5     0.5        0.00           0.0                   0.3   
4         15.0    0.5     0.5        0.00           0.0                   0.3   

   total_amount  
0     17.049999  
1     17.799999  
2     10.800000  
3      4.800000  
4     16.299999

In [53]:
print(ny.shape)
ny.describe()

(499999, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    499999.000000  499999.000000     499999.000000    499999.000000   
mean          1.678109       3.019197        -72.472382        39.960388   
std           1.334983     136.910843         10.150945         5.592647   
min           0.000000       0.000000        -87.451874         0.000000   
25%           1.000000       1.000000        -73.991669        40.735725   
50%           1.000000       1.660000        -73.981567        40.753300   
75%           2.000000       3.000000        -73.966606        40.767715   
max           9.000000   92000.898438          0.000000        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      499999.000000     499999.000000  499999.000000  499999.000000   
mean          -72.518303         39.985832      11.872193       0.314752   
std             9.991468          5.503415      10.134053       0.367454   
min           -86.731705          0.000000    -138.899994      -1.000000   
25%           -73.991219         40.734528       6.500000       0.000000   
50%           -73.979820         40.753757       9.000000       0.000000   
75%           -73.962524         40.768940      13.500000       0.500000   
max             0.000000         49.194656     900.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  499999.000000  499999.000000  499999.000000  
mean        1.542951       0.241650      14.805460  
std         2.513547       1.224226      12.301260  
min       -81.000000      -5.330000    -139.699997  
25%         0.000000       0.000000       8.160000  
50%         1.000000       0.000000      11.150000  
75%         2.060000       0.000000      16.299999  
max       750.000000      95.330002     900.299988

It seems there are negative values in values related to charges (total_amount, extra...).

In [54]:
idx = ny['total_amount'] < 0
print(ny[idx].shape)
ny[idx].head(10)

(158, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
3831         2  2015-01-17 22:40:27   2015-01-17 22:43:04                1   
4924         2  2015-01-15 17:33:24   2015-01-15 17:33:31                2   
10046        2  2015-01-16 16:00:45   2015-01-16 16:00:53                1   
16703        2  2015-01-31 23:38:52   2015-01-31 23:38:54                2   
19953        2  2015-01-10 02:23:53   2015-01-10 02:23:58                2   
22989        2  2015-01-14 11:52:09   2015-01-14 11:52:20                1   
23879        2  2015-01-03 02:01:25   2015-01-03 02:01:54                1   
26582        2  2015-01-12 15:07:29   2015-01-12 15:07:35                1   
27172        2  2015-01-06 14:07:25   2015-01-06 14:08:27                1   
30188        2  2015-01-10 21:10:20   2015-01-10 21:12:39                1   

       trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
3831            0.11        -74.002357        40.739826          1   
4924            0.00        -73.982567        40.739799          1   
10046           0.00        -73.937721        40.758194          1   
16703           0.00          0.000000         0.000000          2   
19953           0.00          0.000000         0.000000          5   
22989           0.00        -73.789955        40.646946          2   
23879           0.03        -73.953407        40.811150          1   
26582           0.00          0.000000         0.000000          2   
27172           0.03        -73.994560        40.740318          1   
30188           0.03        -73.986328        40.755280          1   

      store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
3831                   N         -74.001114         40.741108            4   
4924                   N         -73.982567         40.739799            3   
10046                  N         -73.937721         40.758194            3   
16703                  N           0.000000          0.000000            2   
19953                  N           0.000000          0.000000            1   
22989                  N           0.000000          0.000000            3   
23879                  N         -73.953751         40.811302            2   
26582                  N           0.000000          0.000000            2   
27172                  N         -73.995331         40.740952            4   
30188                  N         -73.985420         40.755089            4   

       fare_amount  extra mta_tax  tip_amount  tolls_amount  \
3831          -3.5   -0.5    -0.5        0.00          0.00   
4924          -2.5   -1.0    -0.5       -0.70          0.00   
10046         -2.5   -1.0    -0.5        0.00          0.00   
16703        -52.0    0.0    -0.5        0.00          0.00   
19953         -6.8    0.0       0       -1.00          0.00   
22989        -52.0    0.0    -0.5      -14.33         -5.33   
23879         -2.5   -0.5    -0.5        0.00          0.00   
26582        -52.0    0.0    -0.5        0.00          0.00   
27172         -2.5    0.0    -0.5        0.00          0.00   
30188         -3.5   -0.5    -0.5        0.00          0.00   

      improvement_surcharge  total_amount  
3831                    0.3     -4.800000  
4924                    0.3     -5.000000  
10046                   0.3     -4.300000  
16703                   0.3    -52.799999  
19953                   0.3     -8.100000  
22989                   0.3    -72.459999  
23879                   0.3     -3.800000  
26582                   0.3    -52.799999  
27172                   0.3     -3.300000  
30188                   0.3     -4.800000

It doesn't make sense unless they are refunds. In that case, it means there are 2 records for the same pickup, one for the original trip and one for the refund.

The data dictionary [https://storage.googleapis.com/hiring-test/data_dictionary_trip_records_yellow.pdf](link) defines the following categories of payment in the payment_type variable:

 - 1= Credit card 
 - 2= Cash
 - 3= No charge 
 - 4= Dispute
 - 5= Unknown 
 - 6= Voided trip

But from the small sample above it seems clear not all refunds are registered as 'Dispute', 'Voided trip'... So this field can't be trusted to identify refunds.

Moving forward, records with a negative total_amount should be discarded to avoid duplicity.


In [55]:
ny = ny[ny['total_amount'] >= 0]
print(ny.shape)
ny.describe()

(499841, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    499841.000000  499841.000000     499841.000000    499841.000000   
mean          1.678048       3.020076        -72.475700        39.962208   
std           1.334955     136.932465         10.140511         5.586823   
min           0.000000       0.000000        -87.451874         0.000000   
25%           1.000000       1.000000        -73.991669        40.735729   
50%           1.000000       1.660000        -73.981567        40.753304   
75%           2.000000       3.000000        -73.966614        40.767715   
max           9.000000   92000.898438          0.000000        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      499841.000000     499841.000000  499841.000000  499841.000000   
mean          -72.522324         39.988068      11.879146       0.314949   
std             9.978506          5.495912      10.123908       0.367364   
min           -86.731705          0.000000       0.000000       0.000000   
25%           -73.991219         40.734531       6.500000       0.000000   
50%           -73.979820         40.753761       9.000000       0.000000   
75%           -73.962532         40.768944      13.500000       0.500000   
max             0.000000         49.194656     900.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  499841.000000  499841.000000  499841.000000  
mean        1.543758       0.241747      14.814079  
std         2.509445       1.224175      12.286285  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.160000  
50%         1.000000       0.000000      11.150000  
75%         2.060000       0.000000      16.299999  
max       750.000000      95.330002     900.299988

It also stands out that the minimum latitude for the pickups and the dropoffs is 0 (pretty far from NYC). The same happens with the max longitude.

In [56]:
idx = ny['pickup_longitude'] == 0
print(ny[idx].shape)
ny[idx].head()

(9537, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
31         2  2015-01-15 19:05:43   2015-01-15 19:05:44                2   
61         1  2015-01-04 13:44:52   2015-01-04 13:56:49                1   
66         2  2015-01-04 13:44:52   2015-01-04 13:49:03                1   
157        1  2015-01-15 09:47:00   2015-01-15 10:00:07                1   
159        1  2015-01-15 09:47:02   2015-01-15 10:17:47                3   

     trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
31            0.01               0.0              0.0          5   
61            2.50               0.0              0.0          1   
66            0.85               0.0              0.0          1   
157           1.00               0.0              0.0          1   
159           8.30               0.0              0.0          1   

    store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
31                   N                0.0               0.0            1   
61                   N                0.0               0.0            1   
66                   N                0.0               0.0            2   
157                  N                0.0               0.0            2   
159                  N                0.0               0.0            1   

     fare_amount  extra mta_tax  tip_amount  tolls_amount  \
31          60.0    0.0       0        0.00          0.00   
61          11.0    0.0     0.5        2.35          0.00   
66           5.5    0.0     0.5        0.00          0.00   
157         10.0    0.0     0.5        0.00          0.00   
159         27.5    0.0     0.5       10.00          5.33   

    improvement_surcharge  total_amount  
31                    0.3     60.299999  
61                      0     14.150000  
66                    0.3      6.300000  
157                   0.3     10.800000  
159                   0.3     43.630001

In [59]:
ny = ny[ny['trip_distance'] != 0]
print(ny.shape)
ny.describe()

(487557, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    487557.000000  487557.000000     487557.000000    487557.000000   
mean          1.685159       3.045619        -73.893021        40.743725   
std           1.342100     138.646118          0.092528         0.075619   
min           0.000000       0.010000        -87.451874         4.789132   
25%           1.000000       1.000000        -73.991859        40.737282   
50%           1.000000       1.680000        -73.981918        40.754162   
75%           2.000000       3.000000        -73.967941        40.768108   
max           6.000000   92000.898438        -69.598526        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      487557.000000     487557.000000  487557.000000  487557.000000   
mean          -73.893669         40.744274      11.834699       0.315770   
std             0.140841          0.048933       9.713044       0.368701   
min           -86.731705         18.625944       0.000000       0.000000   
25%           -73.991379         40.736206       6.500000       0.000000   
50%           -73.980286         40.754509       9.000000       0.000000   
75%           -73.964012         40.769390      13.500000       0.500000   
max            -0.116667         49.194656     420.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  487557.000000  487557.000000  487557.000000  
mean        1.536876       0.240548      14.760816  
std         2.216422       1.220723      11.854428  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.300000  
50%         1.000000       0.000000      11.160000  
75%         2.060000       0.000000      16.299999  
max       200.000000      95.330002     453.299988

It's clear that the location data is invalid for these records. This data will get discarded later when joining with the geojson data. For now I'll discard manually

In [57]:
ny = ny[ny['pickup_longitude'] != 0]
ny = ny[ny['dropoff_longitude'] != 0]
print(ny.shape)
ny.describe()

(489492, 19)


passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
count    489492.000000  489492.000000     489492.000000    489492.000000   
mean          1.683772       3.033579        -73.887733        40.740753   
std           1.341065     138.371902          0.097233         0.075878   
min           0.000000       0.000000        -87.451874         4.789132   
25%           1.000000       1.000000        -73.991852        40.737263   
50%           1.000000       1.670000        -73.981911        40.754145   
75%           2.000000       3.000000        -73.967886        40.768101   
max           9.000000   92000.898438        -69.598526        42.294155   

       dropoff_longitude  dropoff_latitude    fare_amount          extra  \
count      489492.000000     489492.000000  489492.000000  489492.000000   
mean          -73.888496         40.741306      11.856030       0.315313   
std             0.143711          0.049497       9.942463       0.367270   
min           -86.731705         18.625944       0.000000       0.000000   
25%           -73.991379         40.736179       6.500000       0.000000   
50%           -73.980270         40.754494       9.000000       0.000000   
75%           -73.963936         40.769379      13.500000       0.500000   
max            -0.116667         49.194656     900.000000       7.000000   

          tip_amount   tolls_amount   total_amount  
count  489492.000000  489492.000000  489492.000000  
mean        1.541835       0.241740      14.787927  
std         2.493864       1.224812      12.123513  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       8.300000  
50%         1.000000       0.000000      11.150000  
75%         2.060000       0.000000      16.299999  
max       750.000000      95.330002     900.299988

In [58]:
print(ny[ny['trip_distance'] == 0].shape)
ny[ny['trip_distance'] == 0].head()

(1935, 19)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
132        1  2015-01-15 10:26:16   2015-01-15 10:26:42                2   
133        1  2015-01-15 10:26:17   2015-01-15 10:27:28                1   
707        1  2015-01-10 20:14:23   2015-01-10 20:14:36                1   
876        1  2015-01-23 17:43:08   2015-01-23 17:43:31                1   
941        1  2015-01-28 20:22:17   2015-01-28 20:23:36                1   

     trip_distance  pickup_longitude  pickup_latitude RateCodeID  \
132            0.0        -73.987656        40.743645          5   
133            0.0        -73.946526        40.744991          1   
707            0.0        -73.993271        40.748833          1   
876            0.0        -73.970345        40.762089          1   
941            0.0        -73.989746        40.735394          1   

    store_and_fwd_flag  dropoff_longitude  dropoff_latitude payment_type  \
132                  N         -73.987488         40.743484            1   
133                  N         -73.946526         40.744991            2   
707                  N         -73.993271         40.748833            3   
876                  N         -73.970345         40.762089            3   
941                  N         -73.989540         40.735474            3   

     fare_amount  extra mta_tax  tip_amount  tolls_amount  \
132         60.0    0.0       0        15.0           0.0   
133          3.0    0.0     0.5         0.0           0.0   
707          2.5    0.5     0.5         0.0           0.0   
876          2.5    1.0     0.5         0.0           0.0   
941          3.0    0.5     0.5         0.0           0.0   

    improvement_surcharge  total_amount  
132                   0.3     75.300003  
133                   0.3      3.800000  
707                   0.3      3.800000  
876                   0.3      4.300000  
941                   0.3      4.300000

Either the pickup or the dropoff is wrong. Or maybe these records are additional charges made at the dropoff location after a wrong mischarge. If that's the case this is not actually a pickup, so I'm inclinded to delete this records (although I recognize an argument could be made towards keeping them).

There still seem to be wrong coordinates (max dropoff longitude = -0.11) but as I said, they will be discarded later.

There are also lines where the trip distance is 0.

Finally, there are rows where the passenger count is 0, but this can be attributed to a wrognful registry of the data and it doesn't really affect the scope of this analysis.

We started with 499999 records and after cleaning we have 487557. In the next step, we'll try to clean the whole dataset.

# Preprocessing

There is an estimate of 40M records, so I created a helper script (src/read_nyc_data.py) to read only the data I need and store in the feather format. Two files were created, one that only has the columns I need going forward and another with all the columns (in case I need it later)